Can't create custom meta annotation for `@QuarkusTest`

### Describe the bug

I'm trying to compose my own meta annotation that encompasses `@QuarkusTest` as well as additional functionality.

I initially discussed it here: https://quarkusio.zulipchat.com/#narrow/channel/187038-dev/topic/.40QuarkusTest.20annotation.20not.20inheritable.3F/with/510278646 but I decided to put together a simple reproducer which demonstrates the problem.

I tried to compose my own annotation 2 different ways:

```java
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.RUNTIME)
@Documented
@Inherited
@QuarkusTest
public @interface MetaAnnotation {
}
```

and

```java
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.RUNTIME)
@Documented
@Inherited
@ExtendWith(QuarkusTestExtension.class)
public @interface MetaAnnotation2 {
}
```

and placing it on a test class.

### Expected behavior

I would expect to be able to create my own custom annotation that could compose `@QuarkusTest`, or even `@ExtendWith(QuarkusTestExtension.class)`.

### Actual behavior

When I try to do that I get these kinds of errors:

```
org.junit.jupiter.api.extension.TestInstantiationException: Failed to create test instance
        at io.quarkus.test.junit.QuarkusTestExtension.initTestState(QuarkusTestExtension.java:767)
        at io.quarkus.test.junit.QuarkusTestExtension.interceptTestClassConstructor(QuarkusTestExtension.java:733)
        at java.base/java.util.Optional.orElseGet(Optional.java:364)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
        at io.quarkus.runner.bootstrap.RunningQuarkusApplicationImpl.instance(RunningQuarkusApplicationImpl.java:90)
        at io.quarkus.test.junit.QuarkusTestExtension.createActualTestInstance(QuarkusTestExtension.java:773)
        at io.quarkus.test.junit.QuarkusTestExtension.initTestState(QuarkusTestExtension.java:761)
        ... 4 more
Caused by: java.lang.reflect.InvocationTargetException
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at io.quarkus.runner.bootstrap.RunningQuarkusApplicationImpl.instance(RunningQuarkusApplicationImpl.java:88)
        ... 6 more
Caused by: jakarta.enterprise.inject.UnsatisfiedResolutionException: No bean found for required type [class org.example.ExampleResource2Test] and qualifiers [[]]
        at io.quarkus.arc.impl.InstanceImpl.bean(InstanceImpl.java:288)
        at io.quarkus.arc.impl.InstanceImpl.getInternal(InstanceImpl.java:309)
        at io.quarkus.arc.impl.InstanceImpl.get(InstanceImpl.java:190)
        ... 8 more

2025-04-07 12:33:18,765 INFO  [io.quarkus] (main) demo stopped in 0.007s
```

### How to Reproduce?

Reproducer: [demo.zip](https://github.com/user-attachments/files/19636034/demo.zip)

1. Unzip the reproducer
2. `cd demo`
3. `./mvnw clean test`

You'll notice that the `NoMetaAnnotationTest` class runs successfully, yet `MetaAnnotationTest` and `MetaAnnotation2Test` do not.

### Output of `uname -a` or `ver`

Darwin edeandrea-m4max 24.4.0 Darwin Kernel Version 24.4.0: Wed Mar 19 21:17:35 PDT 2025; root:xnu-11417.101.15~1/RELEASE_ARM64_T6041 arm64

### Output of `java -version`

_No response_

### Quarkus version or git rev

3.21.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/docs/src/main/asciidoc/getting-started-testing.adoc
+++ b/docs/src/main/asciidoc/getting-started-testing.adoc
@@ -1640,6 +1640,36 @@ For `@QuarkusIntegrationTest` tests that result in launcher the application as a
 This can be used by `QuarkusTestResourceLifecycleManager` that need to launch additional containers that the application will communicate with.
 ====
 
+== Creating `@QuarkusTest` meta-annotation
+
+The `@QuarkusTest` annotation can be grouped with other annotations by defining a custom meta-annotation.
+For instance, you can bundle it with the `@Tag` annotation to automatically mark a group of tests like in the example below:
+
+[source,java]
+----
+org.acme.test
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Inherited;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import io.quarkus.test.junit.QuarkusTest;
+import org.junit.jupiter.api.Tag;
+
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@Inherited
+@QuarkusTest
+@Tag("security")
+public @interface QuarkusSecurityTest {   <1>
+}
+----
+<1> The `@QuarkusSecurityTest` annotation can be used instead of the `@QuarkusTest` annotation.
+
 == Testing Components
 
 Quarkus provides the `QuarkusComponentTestExtension`, a JUnit extension to ease the testing of components and mocking of their dependencies.

--- a/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/TestsAsBeansProcessor.java
+++ b/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/TestsAsBeansProcessor.java
@@ -5,6 +5,7 @@ import java.util.HashSet;
 import java.util.List;
 import java.util.Optional;
 import java.util.Set;
+import java.util.stream.Collectors;
 
 import org.jboss.jandex.AnnotationInstance;
 import org.jboss.jandex.AnnotationTarget;
@@ -61,7 +62,11 @@ public class TestsAsBeansProcessor {
 
     @BuildStep(onlyIf = IsTest.class)
     AnnotationsTransformerBuildItem vetoTestClassesNotMatchingTestProfile(Optional<TestProfileBuildItem> testProfile,
-            CombinedIndexBuildItem index) {
+            CombinedIndexBuildItem index, List<TestAnnotationBuildItem> testAnnotationBuildItems) {
+        Set<String> additionalTestAnnotationNames = testAnnotationBuildItems.stream()
+                .map(TestAnnotationBuildItem::getAnnotationClassName)
+                .filter(a -> !a.equals(QUARKUS_TEST.toString()))
+                .collect(Collectors.toSet());
         // Since we only need to register all test classes that belong to the "current" test profile
         // we need to veto other test classes during the build
         return new AnnotationsTransformerBuildItem(AnnotationTransformation
@@ -71,7 +76,8 @@ public class TestsAsBeansProcessor {
                     boolean veto = false;
                     if (maybeTestClass.hasAnnotation(QUARKUS_TEST)
                             || maybeTestClass.hasAnnotation(QUARKUS_INTEGRATION_TEST)
-                            || maybeTestClass.hasAnnotation(QUARKUS_MAIN_TEST)) {
+                            || maybeTestClass.hasAnnotation(QUARKUS_MAIN_TEST)
+                            || additionalTestAnnotationNames.stream().anyMatch(maybeTestClass::hasAnnotation)) {
                         String testProfileClassName = testProfile.map(TestProfileBuildItem::getTestProfileClassName)
                                 .orElse(null);
                         veto = !matchesProfile(maybeTestClass, testProfileClassName);

--- a/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/AbstractQuarkusTestMetaAnnotationTest.java
+++ b/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/AbstractQuarkusTestMetaAnnotationTest.java
@@ -1,0 +1,28 @@
+package io.quarkus.it.extension;
+
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.inject.Inject;
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+public abstract class AbstractQuarkusTestMetaAnnotationTest {
+
+    @Inject
+    MyTestBean bean;
+
+    @Test
+    void testInjectedBean() {
+        Assertions.assertEquals("foo", bean.foo());
+    }
+
+    @ApplicationScoped
+    public static class MyTestBean {
+
+        public String foo() {
+            return "foo";
+        }
+
+    }
+
+}

--- a/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/QuarkusTestExtensionMetaAnnotationHierarchyDepth1Test.java
+++ b/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/QuarkusTestExtensionMetaAnnotationHierarchyDepth1Test.java
@@ -1,0 +1,6 @@
+package io.quarkus.it.extension;
+
+@QuarkusTestExtensionMetaAnnotationL1
+public class QuarkusTestExtensionMetaAnnotationHierarchyDepth1Test extends AbstractQuarkusTestMetaAnnotationTest {
+
+}

--- a/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/QuarkusTestExtensionMetaAnnotationHierarchyDepth2Test.java
+++ b/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/QuarkusTestExtensionMetaAnnotationHierarchyDepth2Test.java
@@ -1,0 +1,6 @@
+package io.quarkus.it.extension;
+
+@QuarkusTestExtensionMetaAnnotationL2
+public class QuarkusTestExtensionMetaAnnotationHierarchyDepth2Test extends AbstractQuarkusTestMetaAnnotationTest {
+
+}

--- a/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/QuarkusTestExtensionMetaAnnotationL1.java
+++ b/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/QuarkusTestExtensionMetaAnnotationL1.java
@@ -1,0 +1,20 @@
+package io.quarkus.it.extension;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Inherited;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import org.junit.jupiter.api.extension.ExtendWith;
+
+import io.quarkus.test.junit.QuarkusTestExtension;
+
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@Inherited
+@ExtendWith(QuarkusTestExtension.class)
+public @interface QuarkusTestExtensionMetaAnnotationL1 {
+}

--- a/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/QuarkusTestExtensionMetaAnnotationL2.java
+++ b/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/QuarkusTestExtensionMetaAnnotationL2.java
@@ -1,0 +1,16 @@
+package io.quarkus.it.extension;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Inherited;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@Inherited
+@QuarkusTestExtensionMetaAnnotationL2Parent
+public @interface QuarkusTestExtensionMetaAnnotationL2 {
+}

--- a/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/QuarkusTestExtensionMetaAnnotationL2Parent.java
+++ b/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/QuarkusTestExtensionMetaAnnotationL2Parent.java
@@ -1,0 +1,20 @@
+package io.quarkus.it.extension;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Inherited;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import org.junit.jupiter.api.extension.ExtendWith;
+
+import io.quarkus.test.junit.QuarkusTestExtension;
+
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@Inherited
+@ExtendWith(QuarkusTestExtension.class)
+public @interface QuarkusTestExtensionMetaAnnotationL2Parent {
+}

--- a/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/QuarkusTestMetaAnnotationHierarchyDepth1Test.java
+++ b/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/QuarkusTestMetaAnnotationHierarchyDepth1Test.java
@@ -1,0 +1,6 @@
+package io.quarkus.it.extension;
+
+@QuarkusTestMetaAnnotationL1
+public class QuarkusTestMetaAnnotationHierarchyDepth1Test extends AbstractQuarkusTestMetaAnnotationTest {
+
+}

--- a/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/QuarkusTestMetaAnnotationHierarchyDepth2Test.java
+++ b/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/QuarkusTestMetaAnnotationHierarchyDepth2Test.java
@@ -1,0 +1,6 @@
+package io.quarkus.it.extension;
+
+@QuarkusTestMetaAnnotationL2
+public class QuarkusTestMetaAnnotationHierarchyDepth2Test extends AbstractQuarkusTestMetaAnnotationTest {
+
+}

--- a/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/QuarkusTestMetaAnnotationL1.java
+++ b/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/QuarkusTestMetaAnnotationL1.java
@@ -1,0 +1,21 @@
+package io.quarkus.it.extension;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Inherited;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import org.junit.jupiter.api.Tag;
+
+import io.quarkus.test.junit.QuarkusTest;
+
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@Inherited
+@QuarkusTest
+@Tag("custom-tag")
+public @interface QuarkusTestMetaAnnotationL1 {
+}

--- a/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/QuarkusTestMetaAnnotationL2.java
+++ b/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/QuarkusTestMetaAnnotationL2.java
@@ -1,0 +1,16 @@
+package io.quarkus.it.extension;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Inherited;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@Inherited
+@QuarkusTestMetaAnnotationL2Parent
+public @interface QuarkusTestMetaAnnotationL2 {
+}

--- a/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/QuarkusTestMetaAnnotationL2Parent.java
+++ b/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/QuarkusTestMetaAnnotationL2Parent.java
@@ -1,0 +1,18 @@
+package io.quarkus.it.extension;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Inherited;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import io.quarkus.test.junit.QuarkusTest;
+
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@Inherited
+@QuarkusTest
+public @interface QuarkusTestMetaAnnotationL2Parent {
+}

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/TestBuildChainFunction.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/TestBuildChainFunction.java
@@ -2,8 +2,12 @@ package io.quarkus.test.junit;
 
 import java.nio.file.Path;
 import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 import java.util.ServiceLoader;
 import java.util.function.Consumer;
 import java.util.function.Function;
@@ -15,6 +19,7 @@ import org.jboss.jandex.ClassInfo;
 import org.jboss.jandex.FieldInfo;
 import org.jboss.jandex.Index;
 import org.jboss.jandex.Type;
+import org.junit.jupiter.api.extension.ExtendWith;
 
 import io.quarkus.bootstrap.classloading.QuarkusClassLoader;
 import io.quarkus.builder.BuildChainBuilder;
@@ -31,12 +36,16 @@ import io.quarkus.test.junit.buildchain.TestBuildChainCustomizerProducer;
 
 public class TestBuildChainFunction implements Function<Map<String, Object>, List<Consumer<BuildChainBuilder>>> {
 
+    private static final String QUARKUS_TEST_NAME = QuarkusTest.class.getName();
+    private static final String QUARKUS_TEST_EXTENSION_NAME = QuarkusTestExtension.class.getName();
+
     @Override
     public List<Consumer<BuildChainBuilder>> apply(Map<String, Object> stringObjectMap) {
         Path testLocation = (Path) stringObjectMap.get(AbstractJvmQuarkusTestExtension.TEST_LOCATION);
         // the index was written by the extension
         Index testClassesIndex = TestClassIndexer.readIndex(testLocation,
                 (Class<?>) stringObjectMap.get(AbstractJvmQuarkusTestExtension.TEST_CLASS));
+        Collection<TestAnnotationBuildItem> testAnnotationBuildItems = collectTestAnnotationItems(testClassesIndex);
 
         List<Consumer<BuildChainBuilder>> allCustomizers = new ArrayList<>(1);
         Consumer<BuildChainBuilder> defaultCustomizer = new Consumer<BuildChainBuilder>() {
@@ -76,7 +85,7 @@ public class TestBuildChainFunction implements Function<Map<String, Object>, Lis
                 buildChainBuilder.addBuildStep(new BuildStep() {
                     @Override
                     public void execute(BuildContext context) {
-                        context.produce(new TestAnnotationBuildItem(QuarkusTest.class.getName()));
+                        testAnnotationBuildItems.forEach(context::produce);
                     }
                 })
                         .produces(TestAnnotationBuildItem.class)
@@ -159,5 +168,41 @@ public class TestBuildChainFunction implements Function<Map<String, Object>, Lis
         }
 
         return allCustomizers;
+    }
+
+    private static Collection<TestAnnotationBuildItem> collectTestAnnotationItems(Index testClassesIndex) {
+        var result = new HashSet<String>();
+        result.add(QUARKUS_TEST_NAME);
+
+        // collect meta-annotations with @QuarkusTest
+        result.addAll(collectMetaAnnotations(QUARKUS_TEST_NAME, testClassesIndex));
+        // collect meta-annotations with @ExtendWith(QuarkusTestExtension.class)
+        Predicate<AnnotationInstance> isValueQuarkusTestExtension = ai -> ai.value() != null
+                && Arrays.stream(ai.value().asClassArray())
+                        .anyMatch(type -> type.name() != null && QUARKUS_TEST_EXTENSION_NAME.equals(type.name().toString()));
+        result.addAll(collectMetaAnnotations(ExtendWith.class.getName(), testClassesIndex, isValueQuarkusTestExtension));
+
+        return result.stream().map(TestAnnotationBuildItem::new).toList();
+    }
+
+    private static List<String> collectMetaAnnotations(String annotationName, Index testClassesIndex) {
+        return collectMetaAnnotations(annotationName, testClassesIndex, ai -> true);
+    }
+
+    private static List<String> collectMetaAnnotations(String annotationName, Index testClassesIndex,
+            Predicate<AnnotationInstance> initialFilter) {
+        return testClassesIndex.getAnnotations(annotationName)
+                .stream()
+                .filter(initialFilter)
+                .map(AnnotationInstance::target)
+                .filter(Objects::nonNull)
+                .filter(at -> at.kind() == AnnotationTarget.Kind.CLASS)
+                .map(AnnotationTarget::asClass)
+                .filter(ClassInfo::isAnnotation)
+                .<String> mapMulti((ci, consumer) -> {
+                    consumer.accept(ci.name().toString());
+                    collectMetaAnnotations(ci.name().toString(), testClassesIndex).forEach(consumer);
+                })
+                .toList();
     }
 }

